python/pyspark/util.py

```diff
@@ -263,6 +264,46 @@ def _parse_memory(s):
     return int(float(s[:-1]) * units[s[-1].lower()])
 
 
+def _on_inheritable_thread_start():
+    from pyspark import SparkContext
+    sc = SparkContext._active_spark_context
+    if isinstance(sc._gateway, ClientServer):
+        # Here's when the pinned-thread mode (PYSPARK_PIN_THREAD) is on.
+        properties = sc._jsc.sc().getLocalProperties().clone()
+        sc._jsc.sc().setLocalProperties(properties)
+
+
+def _on_inheritable_thread_stop():
+    from pyspark import SparkContext
+    sc = SparkContext._active_spark_context
+    if isinstance(SparkContext._gateway, ClientServer):
+        thread_connection = sc._jvm._gateway_client.thread_connection.connection()
+        if thread_connection is not None:
+            connections = sc._jvm._gateway_client.deque
+
+            # Reuse the lock for Py4J in PySpark
+            with SparkContext._lock:
+                for i in range(len(connections)):
+                    if connections[i] is thread_connection:
+                        connections[i].close()
+                        del connections[i]
+                        break
+                else:
+                    # Just in case the connection was not closed but removed from the queue.
+                    thread_connection.close()
+
+
+def inheritable_thread(f):
```

Comment:
I would name it a bit differently like `inheritable_thread_target` (given that Python's thread takes `target`) or `inheritable_thread_func` because this function itself isn't a thread.